Use standard select element for small number of authors.

adamsilverstein opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
Use a standard select element when the number of authors falls below a threshold (default is set to 20).

Fixes https://github.com/WordPress/gutenberg/issues/26077

## How has this been tested?

 - Test editing and creating new posts with:
    * default single user: no selector is shown at all.
    * `<` 20 authors: native `<select>` dropdown selector, select, save post.
    * 20 or more authors: ComboboxComponent: search for user, select, save post.

### Helpful commands for a local test environment:

Generate 10 authors: `wp user generate --role=editor --count=10`
Generate 1000 authors: `wp user generate --role=editor --count=1000`
Remove all site editors: `wp user delete $(wp user list --role=editor --field=ID) --reassign=2`

## Screenshots <!-- if applicable -->

Site with 11 editors:
![image](https://user-images.githubusercontent.com/2676022/97031261-39081d00-151d-11eb-8ac0-63a02f8b3212.png)

Site with 100s of editors:
![image](https://user-images.githubusercontent.com/2676022/97031400-7371ba00-151d-11eb-817d-46ce07ceb836.png)





## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

adamsilverstein committed:
Use standard select element for small number of authors.

adamsilverstein commented on packages/editor/src/components/post-author/check.js:
```diff
@@ -40,7 +40,10 @@ export default compose( [
-			authors: select( 'core' ).getUsers( { who: 'authors' } ),
+			authors: select( 'core' ).getUsers( {
+				who: 'authors',
+				per_page: 100,
```
Default in core is now 10? we need to ensure we always get enough to determine if the dropdown should be a combobox or select.

adamsilverstein commented on packages/editor/src/components/post-author/index.js:
```diff
@@ -10,22 +10,32 @@ import { useState, useMemo, useEffect } from '@wordpress/element';
-			const query =
-				! fieldValue || '' === fieldValue ? {} : { search: fieldValue };
+			const isSearching =
+				isCombobox &&
```
Ensure select change doesn't initiate a search.

adamsilverstein commented on packages/editor/src/components/post-author/index.js:
```diff
@@ -39,6 +49,12 @@ function PostAuthor() {
 
+	useEffect( () => {
+		if ( authors ) {
+			setIsCombobox( authors?.length >= minimumUsersForCombobox );
```
Determine if we are displaying a combobox or select (once)

adamsilverstein commented on packages/editor/src/components/post-author/index.js:
```diff
@@ -69,17 +85,21 @@ function PostAuthor() {
 	 */
-	const handleSelect = ( postAuthorId ) => {
-		if ( ! postAuthorId ) {
+	const handleSelect = ( selectedAuthor ) => {
```
Now handles selection from the combobox (the value only) or from the select element (the synthetic event)

adamsilverstein commented on packages/editor/src/components/post-author/index.js:
```diff
@@ -95,6 +115,29 @@ function PostAuthor() {
+				<select
+					id={ selectId }
+					defaultValue={ postAuthor.id }
+					onBlur={ handleSelect }
```
Previously this was `onChange`, linter complained about that not being accessible so switched to `onBlur` and using `defaultValue` vs `value`.

adamsilverstein commented:
@youknowriad this would be nice to get into 5.6 as well, ready for review.

adamsilverstein committed:
merge master

adamsilverstein commented:
@youknowriad refreshed against trunk - this is ready for review.

youknowriad commented on packages/editor/src/components/post-author/index.js:
```diff
@@ -112,4 +153,4 @@ function PostAuthor() {
 }
 
-export default PostAuthor;
+export default compose( [ withInstanceId ] )( PostAuthor );
```
we have a hook now `useInstanceIdd` to replace the HoC

adamsilverstein replied:
👍🏼

youknowriad commented on packages/editor/src/components/post-author/index.js:
```diff
@@ -28,8 +30,14 @@ function PostAuthor() {
+			const isSearching =
+				isCombobox &&
+				( postAuthor
+					? fieldValue && postAuthor.name !== fieldValue
```
Why are we comparing against the  "name" here?

adamsilverstein replied:
This prevent's searching when the author's name matches the input value, the main reason it is here is to prevent a search when initially focusing on the field. Looks like the field now empties when focused, so I can probably remove this.

youknowriad commented on packages/editor/src/components/post-author/index.js:
```diff
@@ -97,6 +115,29 @@ function PostAuthor() {
+		return (
+			<PostAuthorCheck>
+				<label htmlFor={ selectId }>{ __( 'Author' ) }</label>
+				<select
```
label select... are all abstracted in SelectControl, we should probably use that instead

adamsilverstein replied:
Ah, thanks - forgot about that component. I will use it here.

youknowriad commented on packages/editor/src/components/post-author/index.js:
```diff
@@ -97,6 +115,29 @@ function PostAuthor() {
+			</PostAuthorCheck>
+		);
+	}
+
```
I'm not sure I like the fact that both implementations are inside the same component, what if it's instead  two subcomponents and the parent component just switches from one to another depending on the result of `getEntityRecords`?

adamsilverstein replied:
I can give this a try, it might make the code easier to understand.

youknowriad commented on packages/shortcode/src/index.js:
```diff
@@ -90,23 +90,29 @@ export function next( tag, text, index = 0 ) {
+		// Make sure to return any of the extra brackets if they weren't used to
+		// escape the shortcode.
+		return result || result === '' ? left + result + right : match;
+	} );
```
I guess  the changes in this file are  unrelated right?

adamsilverstein replied:
Right! I'll remove these.

youknowriad reviewed (commented):
It's simpler than I thought :) Thanks.

adamsilverstein committed:
Merge branch 'master' into fix/min-users-for-combobox

adamsilverstein committed:
Restore changed file.

adamsilverstein committed:
restore changed file

adamsilverstein committed:
Merge branch 'master' into fix/min-users-for-combobox

adamsilverstein committed:
Rename post-author to post-author-combobox.

adamsilverstein committed:
Reintroduce HTML select based post author selector.

adamsilverstein committed:
Introduce top level PostAuthor displays PostAuthorSelect or PostAuthorCombobox.

adamsilverstein committed:
Reduce post author check query size.

adamsilverstein committed:
Restore file from master.

adamsilverstein commented:
@youknowriad Took another pass at this with the higher level approach you suggested:

* Rename current combobox based `post-author` to `post-author-combobox`, exporting `<PostAuthorCombobox>` component.
* Reintroduce HTML select based post author selector (used previous version, updated to use the new `<SelectControl>` component). In `post-author-select`, export as `<PostAuthorSelect>`
* Introduce top level `PostAuthor` displays `PostAuthorSelect` or `PostAuthorCombobox` based on the number of authors present. With 25 or more available authors, you get the combobox, otherwise you get the standard select (if there is only one author in the system, the selector does not display). 

In addition, I reduced the size of the `PostAuthorCheck` query, we only need to know if there are more than one authors, so I changed the query to a max of 2 users (vs 100), greatly reducing the data sent of the wire for this query (by 50x). Did I miss a reason this should stay at 100?

Could we add `_fields: id` to this query to further reduce data at load? 

On a related note, do our performance metrics measure data over the wire at load? At a minimum it would be worth simulating a system with many users (plus pages and posts) or low low bandwidth for performance checks - then we might see a performance gain (or alternately catch regressions) from this type of change because the page would load faster with less data to load. 

Adding a 'total data loaded' metric would also be interesting, since data can be a metered resource for some users. I would be happy to work on adding this if you agree it would be valuable.

adamsilverstein commented on packages/editor/src/components/post-author/post-author-combobox.js:
```diff
@@ -0,0 +1,108 @@
+/**
```
Note: this file is moved unchanged from packages/editor/src/components/post-author/iundex.js

adamsilverstein commented on packages/editor/src/components/post-author/post-author-select.js:
```diff
@@ -0,0 +1,56 @@
+/**
```
Note: this file is almost unchanged from the packages/editor/src/components/post-author/index.js as it existed prior to introducing the Combobox version, with the once exception that it now uses the `SelectControl` component.

youknowriad commented on packages/editor/src/components/post-author/check.js:
```diff
@@ -40,7 +40,10 @@ export default compose( [
+			authors: select( 'core' ).getUsers( {
+				who: 'authors',
+				per_page: 2,
+			} ),
```
The problem with this change is that it's a different REST API request while `getAuthors` is shared with the other component, which means in one case, we'd do 1 request and the other, we'd do 2. So maybe we should just keep it as is.

adamsilverstein replied:
> while getAuthors is shared with the other component

Out of curiosity, do you know when two requests are the same, is the REST request still made twice?

youknowriad replied:
>  Out of curiosity, do you know when two requests are the same, is the REST request still made twice?

If it's the same "selector" call with the same arguments, it's made once.

adamsilverstein replied:
Reverted changes to packages/editor/src/components/post-author/check.js.

youknowriad commented on packages/editor/src/components/post-author/index.js:
```diff
@@ -1,115 +1,31 @@
 
-export default PostAuthor;
+export default compose( [
+	withSelect( ( select ) => {
```
Should we be using useSelect instead?

adamsilverstein replied:
sure, I'll refactor this.

adamsilverstein replied:
Refactored PostAuthorSelect as a functional component, switching to `useSelect` and `useDispatch`.

youknowriad commented on packages/editor/src/components/post-author/post-author-select.js:
```diff
@@ -0,0 +1,56 @@
+import { decodeEntities } from '@wordpress/html-entities';
+import { SelectControl } from '@wordpress/components';
+
+export class PostAuthorSelect extends Component {
```
Should we make it a functional component instead. It seems that it doesn't really need to be a class.

youknowriad replied:
And maybe take this as an opportunity to use useSelect/useDispatch iinstead.

adamsilverstein replied:
Sure, I brought this in unaltered as it existed previously to make sure it worked as expected. A simple refactor as a functional component would probably make this easier to understand. I'll work on it.

adamsilverstein replied:
Done! Refactored PostAuthorSelect as a functional component, switching to `useSelect` and `useDispatch`.

youknowriad commented:
The current load metric measures the time to load a long post in the editor. Feel free to add new metrics if you think there's any valuable data we can monitor.

adamsilverstein commented:
> The current load metric measures the time to load a long post in the editor. Feel free to add new metrics if you think there's any valuable data we can monitor.

It might be worth measuring a the time to load said post for systems with a large number of users, pages (for page parent selector) and taxonomy terms (eg tags).

Although... since [some of] these load items occur asynchronously and the editor may be usable before they complete, their real world user impact may be limited. I'm curious if the current measurements are for when the editor is ready to accept input or when everything has completely loaded (or both)?

Can you point me to the current load metric tooling and any docs?

adamsilverstein commented:
@youknowriad thank you for the feedback, I will work on addressing these points.

youknowriad commented:
The current metric is for when the block editor is ready to be used (the first block is visible) https://github.com/WordPress/gutenberg/blob/master/packages/e2e-tests/specs/performance/post-editor.test.js#L133-L138

adamsilverstein committed:
Refactor PostAuthorSelect as functional component.

adamsilverstein committed:
Restore packages/editor/src/components/post-author/check.js from master.

adamsilverstein commented:
@youknowriad refactored `PostAuthorSelect` component as functional component, let me know if I missed anything!

youknowriad committed:
Small tweaks

youknowriad commented:
Pushed some small tweaks but this is looking good to me. Thanks Adam.

adamsilverstein commented:
Thanks for the final changes @youknowriad - that looks good!
